lib/hmmlearn/base.py

```diff
@@ -469,32 +471,87 @@ def fit(self, X, lengths=None):
 
         return self
 
+    def _fit_scaling(self, X, stats):
+        frameprob = self._compute_likelihood(X)
+        logprob, fwdlattice, scaling_factors = self._do_forward_scaling_pass(frameprob)
+        bwdlattice = self._do_backward_scaling_pass(frameprob, scaling_factors)
+        posteriors = self._compute_posteriors_scaling(fwdlattice, bwdlattice)
+        self._accumulate_sufficient_statistics_scaling(
+            stats, X, frameprob, posteriors, fwdlattice,
+        bwdlattice)
+        self._accumulate_sufficient_statistics_emissions(
+            stats, X, posteriors, fwdlattice,
+            bwdlattice)
+        return logprob
+
+    def _fit_log(self, X, stats):
+        framelogprob = self._compute_log_likelihood(X)
+        logprob, fwdlattice = self._do_forward_log_pass(framelogprob)
+        bwdlattice = self._do_backward_log_pass(framelogprob)
+        posteriors = self._compute_posteriors(fwdlattice, bwdlattice)
+        self._accumulate_sufficient_statistics(
+            stats, X, framelogprob, posteriors, fwdlattice,
+            bwdlattice)
+        self._accumulate_sufficient_statistics_emissions(
+            stats, X, posteriors, fwdlattice,
+            bwdlattice)
+        return logprob
+
     def _do_viterbi_pass(self, framelogprob):
         n_samples, n_components = framelogprob.shape
         state_sequence, logprob = _hmmc._viterbi(
             n_samples, n_components, log_mask_zero(self.startprob_),
             log_mask_zero(self.transmat_), framelogprob)
         return logprob, state_sequence
 
-    def _do_forward_pass(self, framelogprob):
+    def _do_forward_scaling_pass(self, frameprob):
+        n_samples, n_components = frameprob.shape
+        fwdlattice = np.zeros((n_samples, n_components))
+        scaling_factors = np.zeros(n_samples)
+        success = _hmmc._forward_scaling(n_samples, n_components,
+                                       self.startprob_,
+                                       self.transmat_,
+                                       frameprob,
+                                       fwdlattice,
+                                       scaling_factors)
+        if not success:
+            raise ValueError("Error while scaling, try the log pass")
```

Comment:
Perhaps make the message more explicit ("Consider using implementation='log' instead").

Reply:
:+1: